compiler/rustc_errors/src/markdown.rs

```diff
@@ -0,0 +1,462 @@
+//! A very minimal markdown parser
+//!
+//! Use the entrypoint `create_ast(&str) -> MdTree` to generate the AST.
+#![warn(clippy::pedantic)]
+#![warn(clippy::nursery)]
+
+use regex::bytes::Regex;
+use std::cmp::min;
+use std::io::Error;
+use std::io::Write;
+use std::str::{self, from_utf8};
+use std::sync::LazyLock;
+use termcolor::{Buffer, BufferWriter, Color, ColorChoice, ColorSpec, WriteColor};
+
+const NEWLINE_CHARS: &[u8; 2] = b"\r\n";
+const BREAK_CHARS: &[u8; 10] = br#".,"'\;:?()"#;
+
+/// Representation of how to match various markdown types
+const PATTERNS: [MdPattern; 10] = [
+    MdPattern::new(Anchor::Any("<!--"), Anchor::Any("-->"), MdType::Comment),
+    MdPattern::new(Anchor::Sol("```"), Anchor::Sol("```"), MdType::CodeBlock),
+    MdPattern::new(Anchor::Sol("# "), Anchor::Eol(""), MdType::Heading1),
+    MdPattern::new(Anchor::Sol("## "), Anchor::Eol(""), MdType::Heading2),
+    MdPattern::new(Anchor::Sol("### "), Anchor::Eol(""), MdType::Heading3),
+    MdPattern::new(Anchor::Sol("#### "), Anchor::Eol(""), MdType::Heading4),
+    MdPattern::new(Anchor::LeadBreak("`"), Anchor::TrailBreak("`"), MdType::CodeInline),
+    MdPattern::new(Anchor::LeadBreak("**"), Anchor::TrailBreak("**"), MdType::Strong),
+    MdPattern::new(Anchor::LeadBreak("_"), Anchor::TrailBreak("_"), MdType::Emphasis),
+    MdPattern::new(Anchor::Sol("-"), Anchor::Eol(""), MdType::ListItem),
+    // MdPattern::new(Anchor::Any("\n\n"),Anchor::Any(""))
+    // strikethrough
+];
+
+/// This is an example for using doc comment attributes
+static RE_URL_NAMED: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(
+    r"(?P<pre_ws>^|\s|[[:punct:]])\[(?P<display>.+)\]\((?P<url>\S+)\)(?P<post_ws>$|\s|[[:punct:]])"
+).unwrap()
+});
+static RE_URL: LazyLock<Regex> = LazyLock::new(|| {
+    Regex::new(r"(?P<pre_ws>^|\s|[[:punct:]])<(?P<url>\S+)>(?P<post_ws>$|\s|[[:punct:]])").unwrap()
+});
+
+/// An AST representation of a Markdown document
+#[derive(Debug, PartialEq, Clone)]
+pub enum MdTree<'a> {
+    /// Leaf types
+    Comment(&'a str),
+    CodeBlock(&'a str),
+    CodeInline(&'a str),
+    Strong(&'a str),
+    Emphasis(&'a str),
+    PlainText(&'a str),
+    /// Nexting types
+    Heading1(Vec<MdTree<'a>>),
+    Heading2(Vec<MdTree<'a>>),
+    Heading3(Vec<MdTree<'a>>),
+    Heading4(Vec<MdTree<'a>>),
+    ListItem(Vec<MdTree<'a>>),
+    /// Root node
+    Root(Vec<MdTree<'a>>),
+}
+
+impl<'a> MdTree<'a> {
+    /// Create a `MdTree` from a string and a specified type
+    fn from_type(s: &'a str, tag: MdType) -> Self {
```

Comment:
Instead of having the `MdType` type, `MdPattern` could also contain a function pointer that constructs the `MdTree`. Then the `PATTERNS` constant could just use closures for the non-trivial variants and the tuple variant constructor otherwise